addons/hr_work_entry/models/hr_work_entry.py

```diff
@@ -66,6 +67,52 @@ class HrWorkEntry(models.Model):
     def init(self):
         tools.create_index(self._cr, "hr_work_entry_date_start_date_stop_index", self._table, ["date_start", "date_stop"])
 
+    def _get_actual_durations(self):
+        return sum(work_entry.duration for work_entry in self)
+
+    def _get_calendar(self):
+        return self.employee_id.resource_calendar_id or self.company_id.resource_calendar_id
+
+    def _get_work_entries_data(self, work_entries_dict):
+        data = self.env['hr.work.entry']
+        for key, work_entries in work_entries_dict.items():
+            actual_duration = work_entries._get_actual_durations()
+            week_type = str(key[0].isocalendar()[1] % 2)
+            work_entry_calendar = work_entries._get_calendar()
+            filter_condition = lambda x:\
+                (not work_entry_calendar.two_weeks_calendar or (x.week_type == week_type))\
+                and int(x.dayofweek) == key[0].weekday()\
+                and x.day_period != 'lunch'
+            total_work_entries = work_entry_calendar.attendance_ids.filtered(filter_condition)
+            total_duration = round(sum(work_entry.duration_hours for work_entry in total_work_entries), 2)
```

Comment:
this is a lot of iteration over the `work_entry_calendar.attendance_ids`. Maybe instead of doing so, we could have a map. And the map should probably look like this - 'workday': duration.  It will look different for one week or two week resource calendar though.
If we compute that dictionary before the loop, then inside loop we can just access it directly and won't be that much iteration.
Basically, we should not be iterating over `attendance_ids` in the loop. It's avoidable

Reply:
I've separated the calculation of weekday durations and used them in the work entry